### PR TITLE
add negative shape check to inputLayer

### PR DESCRIPTION
**Pull request by skaae**
Fix https://github.com/Lasagne/Lasagne/issues/457

Add test for negative shape in input Layer. 
The `str(name)` is a bit ugly but the inputLayer might have a `None` name as far as i read the code?

### CODE REVIEW

File: lasagne/layers/input.py [hunk 1]
```diff
@@ -49,6 +49,12 @@ class InputLayer(Layer):
     """
     def __init__(self, shape, input_var=None, name=None, **kwargs):
         self.shape = shape
+        if any(d is not None and d <= 0 for d in self.shape):
+            raise ValueError((
+                "Cannot create InputLayer with a non-positive shape "
+                "dimension. shape=%r, self.name=%s") % (
+                    self.shape, str(name)))
```
Comment (f0k) [inline lasagne/layers/input.py]: Can you please just follow the code from the base Layer constructor? It uses `self.name=%r` and uses `name` directly (not wrapped in `str()` -- note that this doesn't make a difference when using `%s` anyway, it implicitly calls `str()` on the argument to be formatted). Cheers!

File: lasagne/tests/layers/test_input.py [hunk 2]
```diff
@@ -31,3 +31,8 @@ def test_bad_shape_fails(self):
 
         with pytest.raises(ValueError):
             InputLayer((3, 2), input_var)
+
+    def test_negative_shape_fails(self):
+        from lasagne.layers.input import InputLayer
+        with pytest.raises(ValueError):
+            InputLayer((3, -2))
```
Comment (f0k) [inline lasagne/tests/layers/test_input.py]: Should check for zeros as well, not just for negative shapes. This test can basically be copy-pasted from `TestLayer.test_nonpositive_input_dims_raises_value_error()` in `test_base.py`.

File: lasagne/layers/input.py (hunk 1 above)
Comment (skaae) [inline lasagne/layers/input.py]: yes. I thought that would fail, but it seems to work fine.

Comment (f0k) [inline lasagne/layers/input.py]: `%r` also has the advantage that you can still distinguish between a layer that has no name and a layer of the name `"None"`. Not that it would matter really, but it reads better.

File: lasagne/tests/layers/test_input.py (hunk 2 above)
Comment (skaae) [inline lasagne/tests/layers/test_input.py]: Looking at the tests. I'm not sure what Mock(...) does?

Comment (f0k) [inline lasagne/tests/layers/test_input.py]: > I'm not sure what Mock(...) does?

https://docs.python.org/3/library/unittest.mock.html

**Comment by f0k** [comment]
Thanks, looks good now! Merging.
